File: app/src/main/java/com/trevorwiebe/trackacow/activities/ManageDrugsActivity.java
package com.trevorwiebe.trackacow.activities;

import android.content.Intent;

import androidx.activity.result.ActivityResult;
import androidx.activity.result.ActivityResultCallback;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;

import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;

import com.trevorwiebe.trackacow.R;
import com.trevorwiebe.trackacow.adapters.ManageDrugRecyclerViewAdapter;
import com.trevorwiebe.trackacow.dataLoaders.main.drug.QueryAllDrugs;
import com.trevorwiebe.trackacow.db.entities.DrugEntity;
import com.trevorwiebe.trackacow.utils.ItemClickListener;

import java.util.ArrayList;

public class ManageDrugsActivity extends AppCompatActivity implements QueryAllDrugs.OnAllDrugsLoaded {

    private ManageDrugRecyclerViewAdapter mManageDrugRecyclerViewAdapter;
    private ArrayList<DrugEntity> mDrugList = new ArrayList<>();

    private RecyclerView mManageDrugRv;
    private TextView mDrugEmptyList;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_manage_drugs);

        mDrugEmptyList = findViewById(R.id.drug_list_empty);

        FloatingActionButton addNewDrugFab = findViewById(R.id.add_new_drug);
        addNewDrugFab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent addNewDrugIntent = new Intent(ManageDrugsActivity.this, AddNewDrugActivity.class);
                startActivity(addNewDrugIntent);
            }
        });

        mManageDrugRv = findViewById(R.id.manage_drug_rv);
        mManageDrugRv.setLayoutManager(new LinearLayoutManager(this));
        mManageDrugRecyclerViewAdapter = new ManageDrugRecyclerViewAdapter(mDrugList, this);
        mManageDrugRv.setAdapter(mManageDrugRecyclerViewAdapter);

        mManageDrugRv.addOnItemTouchListener(new ItemClickListener(this, mManageDrugRv, new ItemClickListener.OnItemClickListener() {
            @Override
            public void onItemClick(View view, int position) {
                openEditDrugActivityForResult(position);
            }

            @Override
            public void onLongItemClick(View view, int position) {

            }
        }));

        new QueryAllDrugs(this).execute(this);
    }

    ActivityResultLauncher<Intent> editDrugActivityForResult = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            new ActivityResultCallback<ActivityResult>() {
                @Override
                public void onActivityResult(ActivityResult result) {
                    new QueryAllDrugs(ManageDrugsActivity.this).execute(ManageDrugsActivity.this);
                }
            }
    );

    private void openEditDrugActivityForResult(int position){
        DrugEntity selectedDrugObject = mDrugList.get(position);
        Intent editDrugIntent = new Intent(ManageDrugsActivity.this, EditDrugActivity.class);
        editDrugIntent.putExtra("drugObject", selectedDrugObject);
        editDrugActivityForResult.launch(editDrugIntent);
    }

    @Override
    public void onAllDrugsLoaded(ArrayList<DrugEntity> drugEntities) {
        mDrugList = drugEntities;
        if(mDrugList.size() == 0){
            // show drug list empty
            mDrugEmptyList.setVisibility(View.VISIBLE);
        }else{
            // hide drug list empty
            mDrugEmptyList.setVisibility(View.INVISIBLE);
        }
        mManageDrugRecyclerViewAdapter.swapData(mDrugList);
    }
}

File: app/src/main/java/com/trevorwiebe/trackacow/presentation/manage_drugs/ManageDrugsActivity.kt
package com.trevorwiebe.trackacow.presentation.manage_drugs

import androidx.appcompat.app.AppCompatActivity
import com.trevorwiebe.trackacow.domain.adapters.ManageDrugRecyclerViewAdapter
import androidx.recyclerview.widget.RecyclerView
import android.widget.TextView
import android.os.Bundle
import com.trevorwiebe.trackacow.R
import android.content.Intent
import android.view.View
import androidx.activity.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.trevorwiebe.trackacow.presentation.add_edit_drug.AddOrEditDrugActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.trevorwiebe.trackacow.domain.utils.ItemClickListener
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.progressindicator.LinearProgressIndicator
import com.trevorwiebe.trackacow.domain.models.drug.DrugModel
import com.trevorwiebe.trackacow.domain.utils.DataSource
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

@AndroidEntryPoint
class ManageDrugsActivity : AppCompatActivity() {

    // TODO: fix bug where when offline, new drugs added don't show up until you navigate back and to

    private var mManageDrugRecyclerViewAdapter: ManageDrugRecyclerViewAdapter? = null
    private var mDrugList: List<DrugModel> = emptyList()

    private val mManageDrugsViewModel: ManageDrugsViewModel by viewModels()

    private lateinit var mProgressBar: LinearProgressIndicator
    private lateinit var mManageDrugRv: RecyclerView
    private lateinit var mDrugEmptyList: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_manage_drugs)

        mProgressBar = findViewById(R.id.manage_drugs_loading)
        mDrugEmptyList = findViewById(R.id.drug_list_empty)

        val addNewDrugFab = findViewById<FloatingActionButton>(R.id.add_new_drug)
        addNewDrugFab.setOnClickListener {
            val addNewDrugIntent =
                Intent(this@ManageDrugsActivity, AddOrEditDrugActivity::class.java)
            startActivity(addNewDrugIntent)
        }

        mManageDrugRv = findViewById(R.id.manage_drug_rv)
        mManageDrugRv.layoutManager = LinearLayoutManager(this)
        mManageDrugRecyclerViewAdapter = ManageDrugRecyclerViewAdapter(mDrugList, this)
        mManageDrugRv.adapter = mManageDrugRecyclerViewAdapter
        mManageDrugRv.addOnItemTouchListener(
            ItemClickListener(
                this,
                mManageDrugRv,
                object : ItemClickListener.OnItemClickListener {
                    override fun onItemClick(view: View, position: Int) {
                        val editDrugIntent =
                            Intent(this@ManageDrugsActivity, AddOrEditDrugActivity::class.java)
                        editDrugIntent.putExtra("drugObject", mDrugList[position])
                        startActivity(editDrugIntent)
                    }

                    override fun onLongItemClick(view: View, position: Int) {}
                })
        )

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                mManageDrugsViewModel.uiState.collect {

                    if (it.isFetchingFromCloud && it.dataSource == DataSource.Local) {
                        mProgressBar.visibility = View.VISIBLE
                    } else {
                        mProgressBar.visibility = View.GONE
                    }

                    mDrugList = it.drugList

                    if (mDrugList.isEmpty()) {
                        // show drug list empty
                        mDrugEmptyList.visibility = View.VISIBLE
                    } else {
                        // hide drug list empty
                        mDrugEmptyList.visibility = View.INVISIBLE
                    }
                    mManageDrugRecyclerViewAdapter!!.swapData(mDrugList)
                }
            }
        }

    }
}